### PR TITLE
[Refactor] StudentQuestionResource and StudentQuestionAnswerResource use ...ElseThrow alternatives

### PR DESCRIPTION
**Pull request by PatZintel**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context

Higher readability and better adherance to the new Coding Guidelines

### Description

Changed all possible Repository calls to their Else Throw counterpart.

### Test Coverage

StudentQuestionResource.java: 96% (only parts of "mayUpdateSudenQuestionElseThrow" arent included, theses cant be accessed execution right now though) 
SudentQuestionAnswerResourse.java: 100%

### Steps for Testing

Code review,
On Testserver: 
- create a Question to an Exercise or a Lecture
- edit both answer and Question
- delete both Answer and Question

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java [hunk 1]
```diff
@@ -80,15 +80,10 @@ public StudentQuestionAnswerResource(StudentQuestionAnswerRepository studentQues
...
-        }
-        Optional<StudentQuestion> optionalStudentQuestion = studentQuestionRepository.findById(studentQuestionAnswer.getQuestion().getId());
-        if (optionalStudentQuestion.isEmpty()) {
-            return ResponseEntity.notFound().build();
-        }
-        if (!optionalStudentQuestion.get().getCourse().getId().equals(courseId)) {
+        Course course = courseRepository.findByIdElseThrow(courseId);
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, null);
+        StudentQuestion optionalStudentQuestion = studentQuestionRepository.findByIdElseThrow(studentQuestionAnswer.getQuestion().getId());
+        if (!optionalStudentQuestion.getCourse().getId().equals(courseId)) {
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:86]:
```suggestion
        StudentQuestion studentQuestion = studentQuestionRepository.findByIdElseThrow(studentQuestionAnswer.getQuestion().getId());
        if (!studentQuestion.getCourse().getId().equals(courseId)) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java [hunk 2]
```diff
@@ -123,20 +118,13 @@ public StudentQuestionAnswerResource(StudentQuestionAnswerRepository studentQues
...
-            return ResponseEntity.ok().headers(HeaderUtil.createEntityUpdateAlert(applicationName, true, ENTITY_NAME, studentQuestionAnswer.getId().toString())).body(result);
-        }
-        else {
+        StudentQuestionAnswer optionalStudentQuestionAnswer = studentQuestionAnswerRepository.findByIdElseThrow(studentQuestionAnswer.getId());
+        if (!optionalStudentQuestionAnswer.getQuestion().getCourse().getId().equals(courseId)) {
             return forbidden();
         }
+        mayUpdateOrDeleteStudentQuestionAnswerElseThrow(optionalStudentQuestionAnswer, user);
+        StudentQuestionAnswer result = studentQuestionAnswerRepository.save(studentQuestionAnswer);
+        return ResponseEntity.ok().headers(HeaderUtil.createEntityUpdateAlert(applicationName, true, ENTITY_NAME, studentQuestionAnswer.getId().toString())).body(result);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:127]: Here we have `studentQuestionAnswer` and `optionalStudentQuestionAnswer` which now is also a `StudentQuestionAnswer `. Can you use better names and clarify why both are needed?

Comment (PatZintel) [reply src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:127]: I will fix the naming, i already changed that once but this got lost in a merge...
I need the optional (or soon existingStudentQuestionAnswer) to check if the course Id given as PathVariable matches the one in the RequestBody studentQuestionAnswer, to make sure the right answer of the right course gets changed

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java [hunk 3]
```diff
@@ -202,22 +176,17 @@ public StudentQuestionResource(StudentQuestionRepository studentQuestionReposito
...
-        if (!authorizationCheckService.isAtLeastStudentInCourse(lecture.get().getCourse(), user)) {
-            return forbidden();
+        if (lecture.getCourse().getId().equals(courseId)) {
+            authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, lecture.getCourse(), user);
+            List<StudentQuestion> studentQuestions = studentQuestionRepository.findStudentQuestionsForLecture(lectureId);
+            hideSensitiveInformation(studentQuestions);
+            return new ResponseEntity<>(studentQuestions, null, HttpStatus.OK);
         }
-        if (!lecture.get().getCourse().getId().equals(courseId)) {
+        else
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java:188]: please use braces here

**Review by krusche — CHANGES_REQUESTED** [review]
Overall the code changes look good. Great refactoring. Could you make sure that both resources, their corresponding services and repositories have a coverage of 100% and post the coverage as part of the PR?

**Review by SimonEntholzer — COMMENTED** [review]
changes look good, the code is definitely more concise this way.

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java [hunk 4]
```diff
@@ -71,20 +71,15 @@ public StudentQuestionResource(StudentQuestionRepository studentQuestionReposito
      * @throws URISyntaxException if the Location URI syntax is incorrect
      */
     @PostMapping("courses/{courseId}/student-questions")
-    @PreAuthorize("hasRole('USER')")
+    @PreAuthorize("hasRole('USER') && #studentQuestion.course.id == #courseId")
```
[the remark this answers is not in the document]
Comment (PatZintel) [reply src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java:74]: Ill remove it though in favor of more precise error codes

**Review by stefanwaldhauser — APPROVED** [review]
Really great refactoring :) Nice

File: src/main/java/de/tum/in/www1/artemis/repository/StudentQuestionAnswerRepository.java [hunk 5]
```diff
@@ -12,4 +13,8 @@
 @Repository
 public interface StudentQuestionAnswerRepository extends JpaRepository<StudentQuestionAnswer, Long> {
 
+    default StudentQuestionAnswer findByIdElseThrow(long studentQuestionId) {
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/repository/StudentQuestionAnswerRepository.java:16]: you could also add `@NotNull` here, as you did in the other repositories

```suggestion

    @NotNull
    default StudentQuestionAnswer findByIdElseThrow(long studentQuestionId) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java [hunk 6]
```diff
@@ -80,16 +79,11 @@ public StudentQuestionAnswerResource(StudentQuestionAnswerRepository studentQues
...
-        if (optionalStudentQuestion.isEmpty()) {
-            return ResponseEntity.notFound().build();
-        }
-        if (!optionalStudentQuestion.get().getCourse().getId().equals(courseId)) {
-            return forbidden();
+        Course course = courseRepository.findByIdElseThrow(courseId);
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, null);
+        StudentQuestion studentQuestion = studentQuestionRepository.findByIdElseThrow(studentQuestionAnswer.getQuestion().getId());
+        if (!studentQuestion.getCourse().getId().equals(courseId)) {
+            return badRequest("courseId", "400", "PathVariable courseId doesnt match courseId of the StudentQuestionAnswer in the body that should be added");
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:86]:
```suggestion
            return badRequest("courseId", "400", "PathVariable courseId doesn't match courseId of the StudentQuestionAnswer in the body that should be added");
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java [hunk 7]
```diff
@@ -123,65 +117,49 @@ public StudentQuestionAnswerResource(StudentQuestionAnswerRepository studentQues
...
-        }
-        if (mayUpdateOrDeleteStudentQuestionAnswer(optionalStudentQuestionAnswer.get(), user)) {
-            StudentQuestionAnswer result = studentQuestionAnswerRepository.save(studentQuestionAnswer);
-            return ResponseEntity.ok().headers(HeaderUtil.createEntityUpdateAlert(applicationName, true, ENTITY_NAME, studentQuestionAnswer.getId().toString())).body(result);
-        }
-        else {
-            return forbidden();
+        StudentQuestionAnswer existingStudentQuestionAnswer = studentQuestionAnswerRepository.findByIdElseThrow(studentQuestionAnswer.getId());
+        if (!existingStudentQuestionAnswer.getQuestion().getCourse().getId().equals(courseId)) {
+            return badRequest("courseId", "400", "PathVariable courseId doesnt match courseId of the StudentQuestionAnswer in the body");
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:122]:
```suggestion
            return badRequest("courseId", "400", "PathVariable courseId doesn't match courseId of the StudentQuestionAnswer in the body");
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java [hunk 8]
```diff
@@ -123,65 +117,49 @@ public StudentQuestionAnswerResource(StudentQuestionAnswerRepository studentQues
...
+        return ResponseEntity.ok().headers(HeaderUtil.createEntityUpdateAlert(applicationName, true, ENTITY_NAME, studentQuestionAnswer.getId().toString())).body(result);
     }
 
     /**
      * GET /courses/{courseId}/question-answers/:id : get the "id" questionAnswer.
      *
      * @param courseId the id of the course the answer belongs to
      * @param id the id of the questionAnswer to retrieve
-     * @return the ResponseEntity with status 200 (OK) and with body the questionAnswer, or with status 404 (Not Found)
+     * @return the ResponseEntity with status 200 (OK) and with body the questionAnswer, with status 400 (Bad Request) if theres inconsistencies within the data requested, or with status 404 (Not Found)
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:134]:
```suggestion
     * @return the ResponseEntity with status 200 (OK) and with body the questionAnswer, with status 400 (Bad Request) if there are inconsistencies within the data requested, or with status 404 (Not Found)
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java [hunk 9]
```diff
@@ -123,65 +117,49 @@ public StudentQuestionAnswerResource(StudentQuestionAnswerRepository studentQues
...
      */
     @GetMapping("courses/{courseId}/student-question-answers/{id}")
     @PreAuthorize("hasRole('INSTRUCTOR')")
     public ResponseEntity<StudentQuestionAnswer> getStudentQuestionAnswer(@PathVariable Long courseId, @PathVariable Long id) {
         log.debug("REST request to get StudentQuestionAnswer : {}", id);
         User user = this.userRepository.getUserWithGroupsAndAuthorities();
         var course = courseRepository.findByIdElseThrow(courseId);
-        if (!this.authorizationCheckService.isAtLeastStudentInCourse(course, user)) {
-            return forbidden();
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, user);
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:142]: you can further refactor the code by setting user to null. It is getting fetched by the repository if it is not available :)

```suggestion
        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, null);
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java [hunk 10]
```diff
@@ -123,65 +117,49 @@ public StudentQuestionAnswerResource(StudentQuestionAnswerRepository studentQues
...
      * @param courseId the id of the course the answer belongs to
      * @param id the id of the questionAnswer to retrieve
-     * @return the ResponseEntity with status 200 (OK) and with body the questionAnswer, or with status 404 (Not Found)
+     * @return the ResponseEntity with status 200 (OK) and with body the questionAnswer, with status 400 (Bad Request) if theres inconsistencies within the data requested, or with status 404 (Not Found)
      */
     @GetMapping("courses/{courseId}/student-question-answers/{id}")
     @PreAuthorize("hasRole('INSTRUCTOR')")
     public ResponseEntity<StudentQuestionAnswer> getStudentQuestionAnswer(@PathVariable Long courseId, @PathVariable Long id) {
         log.debug("REST request to get StudentQuestionAnswer : {}", id);
         User user = this.userRepository.getUserWithGroupsAndAuthorities();
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:140]: remove in favor of refactoring
```suggestion
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java [hunk 11]
```diff
@@ -194,29 +172,24 @@ else if (studentQuestionAnswer.getQuestion().getExercise() != null) {
...
-            log.info("StudentQuestionAnswer deleted by {}. Answer: {} for {}", user.getLogin(), studentQuestionAnswer.getAnswerText(), entity);
-            studentQuestionAnswerRepository.deleteById(id);
-            return ResponseEntity.ok().headers(HeaderUtil.createEntityDeletionAlert(applicationName, true, ENTITY_NAME, id.toString())).build();
-        }
-        else {
-            return forbidden();
+            return badRequest("courseId", "400", "PathVariable courseId doesnt match courseId of the StudentQuestionAnswer that should be deleted");
         }
+        mayUpdateOrDeleteStudentQuestionAnswerElseThrow(studentQuestionAnswer, user);
+        log.info("StudentQuestionAnswer deleted by " + user.getLogin() + ". Answer: " + studentQuestionAnswer.getAnswerText() + " for " + entity, user.getLogin());
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:178]: the second argument is not needed anymore

```suggestion
        log.info("StudentQuestionAnswer deleted by " + user.getLogin() + ". Answer: " + studentQuestionAnswer.getAnswerText() + " for " + entity);
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java [hunk 12]
```diff
@@ -67,24 +67,24 @@ public StudentQuestionResource(StudentQuestionRepository studentQuestionReposito
      *
      * @param courseId course the question belongs to
      * @param studentQuestion the studentQuestion to create
-     * @return the ResponseEntity with status 201 (Created) and with body the new studentQuestion, or with status 400 (Bad Request) if the studentQuestion has already an ID
+     * @return the ResponseEntity with status 201 (Created) and with body the new studentQuestion, or with status 400 (Bad Request) if the studentQuestion
+     * already
+     * has an ID or the courseId in the body doesnt match the PathVariable
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java:72]:
```suggestion
     * already has an ID or the courseId in the body doesn't match the PathVariable
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java [hunk 13]
```diff
@@ -67,24 +67,24 @@ public StudentQuestionResource(StudentQuestionRepository studentQuestionReposito
...
      * @throws URISyntaxException if the Location URI syntax is incorrect
      */
     @PostMapping("courses/{courseId}/student-questions")
     @PreAuthorize("hasRole('USER')")
     public ResponseEntity<StudentQuestion> createStudentQuestion(@PathVariable Long courseId, @RequestBody StudentQuestion studentQuestion) throws URISyntaxException {
+        if (!studentQuestion.getCourse().getId().equals(courseId)) {
+            return badRequest("courseId", "400", "PathVariable courseId doesnt match the courseId of the sent StudentQuestion in Body");
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java:79]: check for other `doesnt` misspellings in your code ;)
```suggestion
            return badRequest("courseId", "400", "PathVariable courseId doesn't match the courseId of the sent StudentQuestion in Body");
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java [hunk 14]
```diff
@@ -231,11 +211,8 @@ public StudentQuestionResource(StudentQuestionRepository studentQuestionReposito
     public ResponseEntity<List<StudentQuestion>> getAllQuestionsForCourse(@PathVariable Long courseId) {
         final User user = userRepository.getUserWithGroupsAndAuthorities();
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java:212]: you can delete User here

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java [hunk 15]
```diff
@@ -273,33 +250,42 @@ else if (studentQuestion.getExercise() != null) {
...
-        if (mayUpdateOrDeleteStudentQuestion(studentQuestion, user)) {
-            log.info("StudentQuestion deleted by {}. Question: {} for {}", user.getLogin(), studentQuestion.getQuestionText(), entity);
-            studentQuestionRepository.deleteById(studentQuestionId);
-            return ResponseEntity.ok().headers(HeaderUtil.createEntityDeletionAlert(applicationName, true, ENTITY_NAME, studentQuestionId.toString())).build();
-        }
-        else {
-            return forbidden();
-        }
+        mayUpdateOrDeleteStudentQuestionElseThrow(studentQuestion, user);
+        log.info("StudentQuestion deleted by " + user.getLogin() + ". Question: " + studentQuestion.getQuestionText() + " for " + entity, user.getLogin());
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java:254]: too many arguments

```suggestion
        log.info("StudentQuestion deleted by " + user.getLogin() + ". Question: " + studentQuestion.getQuestionText() + " for " + entity);
```

**Review by derLalla — CHANGES_REQUESTED** [review]
This is some nice refactoring 👍 
But I have found some smaller issues you can improve :)

**Review by katjanakosic — APPROVED** [review]
Tested on TS1. Everything looks fine.

**Review by ArtemTsurikov — APPROVED** [review]
Tested with @katjanakosic. Looks good!

**Review by NicolasUllmann — COMMENTED** [review]
Tested on Server 2 together with @willergeorg. Looked fine, everything worked.

**Review by willergeorg — COMMENTED** [review]
Tested with @NicolasUllmann on TS2. Seems to work!

**Review by joschahenningsen — APPROVED** [review]

**Pull request merged**